File: haskell/ebs.hs
module Ebs (
    ebs,
    ebsmod
) where

--
-- ebsaux f 1 b n = f b^n
--
ebsaux :: (Int -> Int) -> Int -> Int -> Int -> Int
ebsaux f a b 0 = f a
ebsaux f a b 1 = f (a*b)
ebsaux f a b n =
    let
       a' = case mod n 2 of 1 -> f (a*b)
                            _ -> f a
    in
       ebsaux f a' (f (b*b)) (div n 2)

--
-- ebs x n = x^n
--
ebs :: Int -> Int -> Int
ebs x = ebsaux id 1 x

--
-- ebsmod x n m = x^n (mod m)
--
ebsmod :: Int -> Int -> Int -> Int
ebsmod x n m = ebsaux (`mod` m) 1 x n

File: erlang/ebs.erl
-module(ebs).
-export([ebs/2,ebsmod/3]).

%
%  ebsaux(F,1,X,N) => F(X^N)
%
ebsaux(F,A,_,0) -> F(A);
ebsaux(F,A,B,1) -> F(A * B);
ebsaux(F,A,B,N) ->
    A1 = case N rem 2 of
            1 -> F(A * B);
            _ -> F(A)
         end,
    B1 = F(B * B),
    N1 = N div 2,
    ebsaux(F,A1,B1,N1).


%
%  ebs x n => x^n
%
ebs(X,N) ->
    ID = fun (Y) -> Y end,
    ebsaux(ID, 1, X, N).


%
%  ebsmod(X,N,M) => X^N (mod M)
%
ebsmod(X,N,M) ->
    MOD = fun (A) -> A rem M end,
    ebsaux(MOD,1,X,N).
